### PR TITLE
CASSANDRASC-76: Sidecar does not handle keyspaces and table names wit…

### PR DESCRIPTION
**Pull request by frankgh**
…h mixed case

Cassandra Sidecar does not properly handle API requests for endpoints that have keypaces or table names that need to be quoted, for example when names have mixed case or when the name is a reserved keyword in Cassandra.

In this commit, we perform special handling when the keyspace or table names are quoted in the path params. We add tests to ensure that handling is correct.

Additionally, we fix the validation for keyspaces and table names without quotes and add special validation for quoted names.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -323,6 +323,8 @@ tasks.register("integrationTest", Test) {
         jvmArgs(JDK11_OPTIONS)
         println("JVM arguments for $project.name are $allJvmArgs")
     }
+    // Disable direct memory allocator as it doesn't release properly
+    systemProperty "cassandra.netty_use_heap_allocator", "true"
```
Comment (frankgh) [inline build.gradle:327]: we can now run two integration tests concurrently in CI with this setting

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Keyspace.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.common.data;
+
+/**
+ * Represents a Cassandra keyspace
+ */
+public class Keyspace extends Name
```
Comment (yifan-c) [inline common/src/main/java/org/apache/cassandra/sidecar/common/data/Keyspace.java]: This data class is unnecessary. The keyspace value can use the type `Name` directly.

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Table.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.common.data;
+
+/**
+ * Represents a Cassandra table
+ */
+public class Table extends Name
```
Comment (yifan-c) [inline common/src/main/java/org/apache/cassandra/sidecar/common/data/Table.java]: It is unnecessary too. I think it is valuable to have less code.

**Comment by frankgh** [comment]
I had to rebase and force push to resolve conflicts

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Name.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Constructs a {@link Name} object with the provided {@code unquotedName}, and
+     * {@code isQuotedFromSource} parameters. For example, if the user-provided input was
+     * {@code "\"QuotedName\""}, then the parameters for this constructor should be initialized with
+     * {@code new Name("QuotedTableName", true)}.
+     *
+     * @param unquotedName       the unquoted name for the table
+     * @param isQuotedFromSource whether the name was quoted from source
+     */
+    Name(String unquotedName, boolean isQuotedFromSource)
```
Comment (yifan-c) [inline common/src/main/java/org/apache/cassandra/sidecar/common/data/Name.java]: Should it be named as `isQuotedInSource`? Frankly, the "source" is a bit unclear when reading. The value is from the URL segment. Maybe just `isQuoted` is sufficient for the purpose of `Name`.

File: src/main/java/org/apache/cassandra/sidecar/utils/CassandraInputValidator.java [hunk 5]
```diff
@@ -65,13 +67,16 @@ public CassandraInputValidator()
...
-        validatePattern(keyspace, "keyspace");
-        if (validationConfiguration.forbiddenKeyspaces().contains(keyspace))
+        String unquoted = removeQuotesIfNecessary(keyspace);
+        // whether the input is quoted from the source
+        boolean isQuoted = !unquoted.equals(keyspace);
+        validatePattern(unquoted, keyspace, "keyspace", isQuoted);
+        if (validationConfiguration.forbiddenKeyspaces().contains(unquoted))
             throw new HttpException(HttpResponseStatus.FORBIDDEN.code(), "Forbidden keyspace: " + keyspace);
-        return keyspace;
+        return new Keyspace(unquoted, isQuoted);
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/utils/CassandraInputValidator.java]: The implementation expects the value provided in the url to be explicitly quoted. Getting rid of the quotations and adding them back in `Name` seems unnecessary. 
Just passing the input value as-is (either a quoted string or not) should work the same. Right?

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Keyspace.java (hunk 2 above)
Comment (frankgh) [reply common/src/main/java/org/apache/cassandra/sidecar/common/data/Keyspace.java]: done

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Table.java (hunk 3 above)
Comment (frankgh) [reply common/src/main/java/org/apache/cassandra/sidecar/common/data/Table.java]: done

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/Name.java (hunk 4 above)
Comment (frankgh) [reply common/src/main/java/org/apache/cassandra/sidecar/common/data/Name.java]: we no longer have this constructor, so it's no longer relevant

File: src/main/java/org/apache/cassandra/sidecar/utils/CassandraInputValidator.java (hunk 5 above)
Comment (frankgh) [reply src/main/java/org/apache/cassandra/sidecar/utils/CassandraInputValidator.java]: done

**Comment by frankgh** [comment]
Closed via https://github.com/apache/cassandra-sidecar/commit/31fa33fcb446e522947f899d948de4042be04c62

**Pull request closed without merging**